toolkit/scripts/toolchain.mk

```diff
@@ -139,14 +145,49 @@ $(raw_toolchain): $(toolchain_files)
 			$(SPECS_DIR) \
 			$(SOURCE_URL)
 
-# Always start with a fresh toolchain chroot when rebuilding toolchain RPMs
+# This target establishes a cache of toolchain RPMs for partially rehydrating the toolchain from package repos.
+# $(toolchain_from_repos) is a staging folder for these RPMs. We use the toolchain manifest to get a list of
+# filenames for each possibly-rehydrated RPM. We attempt to download each RPM from $(PACKAGE_URL_LIST). In cases
+# where an RPM is unavailable/download fails, we create an empty file. The build_official_toolchain_rpms.sh script
+# knows the distinction between empty/non-empty files.
 #
+# When is it meaningful to attempt to partially rehydrate? It makes no sense to partially rehydrate if:
+# - REBUILD_TOOLCAHIN = n: We aren't building the toolchain, so we defer to the full rehydration step
+# - INCREMENTAL_TOOLCHAIN = n: We explicitly want to build a full toolchain
+# - ALLOW_TOOLCHAIN_DOWNLOAD_FAIL = n: This flag explicitly disables partial toolchain rehydration from repos
+# In these cases, we just create empty files for each possible rehydrated RPM.
+ifeq ($(strip $(INCREMENTAL_TOOLCHAIN))$(strip $(REBUILD_TOOLCHAIN))$(strip $(ALLOW_TOOLCHAIN_DOWNLOAD_FAIL)),yyy)
+$(toolchain_rpms_rehydrated): $(TOOLCHAIN_MANIFEST)
+	@rpm_filename="$(notdir $@)" && \
+	rpm_dir="$(dir $@)" && \
+	log_file="$(toolchain_downloads_logs_dir)/$$rpm_filename.log" && \
+	echo "Attempting to download toolchain RPM: $$rpm_filename" | tee "$$log_file" && \
+	mkdir -p $$rpm_dir && \
+	cd $$rpm_dir && \
+	for url in $(PACKAGE_URL_LIST); do \
+		wget $$url/$$rpm_filename \
+			$(if $(TLS_CERT),--certificate=$(TLS_CERT)) \
+			$(if $(TLS_KEY),--private-key=$(TLS_KEY)) \
+			-a $$log_file && \
+		echo "Downloaded toolchain RPM: $$rpm_filename" >> $$log_file && \
+		echo "$$rpm_filename" >> $(toolchain_downloads_manifest) | tee "$$log_file" && \
+		touch $@ && \
+		break; \
+	done || { \
+		echo "Could not find toolchain package in package repo: $$rpm_filename." | tee "$$log_file" && \
+		touch $@; \
+	}
+else
+$(toolchain_rpms_rehydrated): $(TOOLCHAIN_MANIFEST) 
```

Comment:
So these empty files will be created even if we are doing a full rebuild of the toolchain?
Do they get cleaned?

Reply:
Yep, these are created even if doing a full rebuild of the toolchain. If that's a concern, I could probably avoid keeping all of these files by having a status flag file indicating "we've attempted to rehydrate, if applicable", and having that flag conditionally depend on `$(toolchain_rpms_rehydrated)`. It would be similar to the pattern we've established with `$(STATUS_FLAGS_DIR)/build_toolchain_srpms.flag`.

These files are cleaned by the `clean-toolchain` target, when the directory containing them is removed. Do you want to see a separate target for these files?

Reply:
Deferring any changes from this comment to a future PR, as it does not affect functionality. Let me know if you want me to make any of the changes listed in my comment.